decider-halting-segment-reproduction/src/main.rs

```diff
@@ -74,63 +78,63 @@ enum HaltingSegmentResult {
     NodeLimitExceeded,
 }
 
-fn get_initial_nodes(tm: &TM, segment_size: u8, initial_pos_in_segment: usize) -> Vec<Node> {
+fn get_initial_nodes(tm: &TM, segment_size: u8, initial_pos_in_segment: usize) -> IndexSet<Node> {
     assert!(initial_pos_in_segment < segment_size as usize);
-    let mut to_return: Vec<Node> = Vec::new();
+    let mut to_return: IndexSet<Node> = IndexSet::new();
     for i_state in 0..tm.n_states {
         for symbol in 0..tm.n_symbols {
             let transition = tm.transitions[i_state as usize][symbol as usize];
             if let HaltOrGoto::Halt = transition.goto {
                 let mut initial_segment = vec![SegmentCell::Unallocated; segment_size as usize];
                 initial_segment[initial_pos_in_segment] = SegmentCell::Bit(u8_to_bool(symbol));
-                to_return.push(Node {
+                to_return.insert(Node {
                     state: OutsideSegmentOrState::State(i_state),
                     segment: SegmentCells(initial_segment),
                     pos_in_segment: initial_pos_in_segment,
-                })
+                });
             }
         }
     }
 
     to_return
 }
 
+enum NodeLimit {
```

Comment:
btw, its ok to be lazy and use `Option<usize>` for such cases. then you can write:
```rust
if node_limit.map(|limit| idx_seen > limit).unwrap_or(false) {
    return HaltingSegmentResult::NodeLimitExceeded;
}
```

Reply:
while looking at it, you can move it after `nodex.extend(...` & compare `nodes.len() > limit` to fail faster.

Reply:
Thank you!!